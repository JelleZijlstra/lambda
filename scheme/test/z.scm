; Recursion using the Z combinator

(define z (lambda (f)
    (let ((g (lambda (x) (f (lambda (v) ((x x) v))))))
        (g g))))

(define length (z (lambda (length) (lambda (lst)
    (if (null? lst)
        0
        (+ 1 (length (cdr lst))))))))

(print (length '(1 2 3)))
(print (length '()))
(print (length '(1)))
