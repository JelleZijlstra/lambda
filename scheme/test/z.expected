%%stdout%%
3
0
1
%%stderr%%
